[SYCL][CUDA] Non-uniform algorithm implementations for ext_oneapi_cuda.

JackAKirk opened the pull request:
This PR adds cuda support for fixed_size_group, ballot_group, and opportunistic_group algorithms. All group algorithm support added for the SPIRV impls (those added in e.g. https://github.com/intel/llvm/pull/9181) is correspondingly added here for the cuda backend.

Everything except the reduce/scans uses the same impl for all non-uniform groups. Reduce algorithms also use the same impl for all group types on sm80 for special IsRedux types/ops pairs.

Otherwise reduce/scans have two impl categories:
1.fixed_size_group
2.opportunistic_group, ballot_group, (and tangle_group once it is supported) all use the same impls.

Note that tangle_group is still not supported. However all algorithms implemented by ballot_group/opportunistic_group will I think be appropriate for tangle_group when it is supported.

Pennycook commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,370 @@
+
+//// Shuffle based masked reduction impls
+
+// Cluster group reduction using shfls, T = double
```
There are a few instances of "Cluster group" here (mostly in comments), presumably predating the name change.  Probably a good idea to change them to "Fixed size group" to avoid any confusion.

Pennycook commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,370 @@
+  if constexpr (ext::oneapi::experimental::is_fixed_size_group_v<Group>) {
+    return __nvvm_shfl_sync_up_i32(MemberMask, x, delta, 0);
+  } else {
+    return __nvvm_shfl_sync_idx_i32(MemberMask, x, delta, 31);
```
Something about this line looks off.  If it's an indexed shuffle, should it be accessing the current lane + delta?

JackAKirk commented on sycl/include/sycl/detail/spirv.hpp:
```diff
@@ -137,26 +141,35 @@ bool GroupAll(ext::oneapi::experimental::ballot_group<ParentGroup> g,
 bool GroupAll(ext::oneapi::experimental::tangle_group<ParentGroup>, bool pred) {
   return __spirv_GroupNonUniformAll(group_scope<ParentGroup>::value, pred);
 }
-template <typename Group>
```
@Pennycook Note that I removed this apparently unused template parameter: just want to check this is OK?

JackAKirk commented:
Looks like header includes may have caused a load of tests to fail. Need to investigate

JackAKirk commented:
> Looks like header includes may have caused a load of tests to fail. Need to investigate

I've fixed this now: the opportunistic_group header I added to fix a weird forward decl issue that only occured in nvptx backend was causing issues. I've moved all nvptx algorithm in spirv.hpp that were split up into fixed_size_group and other group type spirv specialized functions to group_algorithm.hpp to avoid this. Actually this is probably simpler anyway since it avoids nvptx duplications since for all group types the masked implementation of these algorithms is the same.

Ready for review @KseniyaTikhomirova . Thanks

KseniyaTikhomirova reviewed (commented):
@JackAKirk I do not have expertise with the code in ext/oneapi/experimental/cuda, do you know anyone who has? Would be great to ask them to review it

Pennycook commented on sycl/include/sycl/detail/spirv.hpp:
```diff
@@ -1022,8 +1022,10 @@ ControlBarrier(Group, memory_scope FenceScope, memory_order Order) {
+ControlBarrier(Group g, memory_scope FenceScope, memory_order Order) {
+#if defined(__NVPTX__)
+  __nvvm_bar_warp_sync(detail::ExtractMask(detail::GetMask(g))[0]);
+#else
```
I have no preference.  But what you've done here is consistent with other parts of DPC++, at least.  For example, the [sub-group implementation](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/sub_group.hpp#L306-L316) assumes that SPIR-V intrinsics will be supported.  I think this makes sense, because some SPIR-V intrinsics are implemented in libclc.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,367 @@
+  auto opRange = g.get_local_range()[0];
+
+  // remainder that won't have a binary partner each loop iteration
+  int remainder;
```
Is there a reason for this to be declared outside the loop?

JackAKirk replied:
There was originally but I think you are right and I will declare it in the loop.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,367 @@
+      x = binary_op(x, tmp);
+    }
+
+    opRange = std::ceil((float)opRange / 2.0f);
```
Given `remainder` was likely just a pretty cheap bit-check, couldn't we update `opRange` cheaper as `stride + remainder`? The compiler might be able to do it for us, but I am not confident enough that it will.

JackAKirk replied:
I don't think so because `remainder + stride` = `(opRange % 2) + (opRange / 2)` doesn't equal `std::ceil((float)opRange / 2.0f)` generally I think. for example remainder can be more than 1. I'll think on what you say more in case I misunderstood.

steffenlarsen replied:
I won't rule out I am wrong, but how can `remainder` be anything other than `0` or `1`? For any positive integer value `x`, `x % 2` should always be either `0` (`x` is even) or `1` (`x` is odd.) Likewise, you would only have to add 1 to the floored division if `x` was odd, so in my head it works out.

JackAKirk replied:
D'oh. Of course you are right... I'll fix it.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,367 @@
+    remainder = opRange % 2;
+
+    // stride between local_ids forming a binary op
+    int stride = opRange / 2;
```
Nit: Any reason for this (and `unfoldedSrcSetBit`) to be signed?

JackAKirk replied:
No they can be unsigned. I will change this.

JackAKirk replied:
Note however that `unfoldedSrcSetBit` has to be signed in `masked_scan_cuda_shfls`

steffenlarsen replied:
If it makes sense to keep it signed, it's no big deal. 😄

JackAKirk replied:
OK I've kept just `unfoldedSrcSetBit` signed, consistent with the function where this variable name needs to be signed.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,367 @@
+    } else if constexpr (std::is_same_v<T, unsigned int>) {                    \
+      return std::numeric_limits<unsigned int>::OP();                          \
+    } else if constexpr (std::is_same_v<T, long>) {                            \
+      return std::numeric_limits<int>::OP();                                   \
```
Is it intentional that we return the numeric limits of `int` when asked for the ones for `long`?

JackAKirk replied:
I did do this intentionally but I think it is wrong and I will correct it!

steffenlarsen replied:
In that case, is there a reason we can't just have it return `std::numeric_limits<T>::OP();` no matter the type, or are there types outside these checks?

Pennycook commented on sycl/include/sycl/detail/type_traits.hpp:
```diff
@@ -50,6 +50,11 @@ template <class T>
+template <class T> struct is_fixed_size_group : std::false_type {};
+
+template <class T>
+inline constexpr bool is_fixed_size_group_v = is_fixed_size_group<T>::value;
```
I'd prefer it in `detail::`.  I missed this during my review -- I thought that because this header is in `detail`, everything was in the `detail` namespace.  If that's not true, it would be nice if somebody (outside of this PR) could split them out.

JackAKirk commented on sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp:
```diff
@@ -0,0 +1,367 @@
+  } else {
+    auto input = std::is_same_v<T, float> ? __nvvm_bitcast_f2i(x) : x;
+    auto tmp_b32 = non_uniform_shfl_T<Group>(MemberMask, input, shfl_param);
+    res = std::is_same_v<T, float> ? __nvvm_bitcast_i2f(tmp_b32) : tmp_b32;
```
Somehow previously I convinced myself from checking cuda's implementation that long is always treated as 32 bit and that half was not supported. Anyhow you are right and I will add these cases.

JackAKirk commented:
Thanks for the reviews.
Hopefully I've addressed the points now. I've added all missing supported types. The half impl can be made faster but I think the appropriate place to address this is not this PR: see https://github.com/intel/llvm/issues/9809

I've verified all types, but I think I should also really add more test coverage for all supported types. But maybe this should be done in a later PR.

steffenlarsen reviewed (approved):
LGTM! We seem to have testing for most of this, so I am okay with doing a follow-up expanding the coverage.

JackAKirk commented:
> LGTM! We seem to have testing for most of this, so I am okay with doing a follow-up expanding the coverage.

Thanks!

jinz2014 commented:
@JackAKirk 
I have some question. Is your implementation related to masked shuffle functions in CUDA ?  I hope more examples will be available for users.  Thanks.

JackAKirk commented:
> @JackAKirk I have some question. Is your implementation related to masked shuffle functions in CUDA ? I hope more examples will be available for users. Thanks.

Yes masked shuffle functions are used throughout. Note masked shuffle functions are not currently exposed in the extension. I do not know if there is a demand/plan to expose them as part of this sycl extension. I have not come across them very often so far.

jinz2014 commented:
A blog like https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/ for SYCL will be read by many users.
